docs/spec/v1alpha1/receivers.md

```diff
@@ -0,0 +1,52 @@
+# Receivers
+
+The `Receiver` API defines a webhook receiver that triggers
+a synchronization for a group of sources.
+
+## Specification
+
+```go
+type ReceiverSpec struct {
+	// Type of webhook sender, used to determine
+	// the validation procedure and payload deserialization.
+	// +kubebuilder:validation:Enum=github;gitlab
+	// +required
+	Type string `json:"type"`
+
+	// A list of sources to be notified about changes.
+	// +required
+	Sources []corev1.TypedLocalObjectReference `json:"sources"`
+}
+```
+
+## Status
+
+```go
+type ReceiverStatus struct {
+	// Generated webhook URL.
+	// +required
+	URL string `json:"url"`
+
+	// Generate token used to validate the payload authenticity.
+	// +required
+	Token string `json:"token"`
+}
+```
+
+## Implementation
+
+The source controller would expose an HTTP server on dedicated port
+(separate from the static file server) to handle the webhook POST requests.
+
+When a `Receiver` is created, the controller generates a random token and
+sets the `Receiver` status token and URL in the format `/hook/sha256sum(token)`.
+The `ReceiverReconciler` creates an indexer for the SHA265 digest
+so that it can be used as a field selector.
+
+When source controller receives a POST request:
+* extract the SHA265 digest from the URL
+* loads the `Receiver` using the digest field selector
+* extracts the signature from HTTP headers based on `spec.type`
+* validates the signature using `status.Token` based on `spec.type`
+* extract the event type from the payload 
+* triggers a synchronization for `spec.sources` if the event type is `push`
```

Comment:
Not related to the current revision of the spec, but looking at the events GitHub supports we could utilize this to create on-demand `GitRepository` sources using the `pull_request` event.

I do however think this would require a change to the `GitRepository` spec (or an additional custom resource) so that it is possible to create multiple artifact revisions for a single repository / credential set.